crawl-ref/source/webserver/ws_handler.py

```diff
@@ -392,13 +392,17 @@ def do_login(self, username):
             self.send_game_links()
 
     def login(self, username, password):
-        real_username = userdb.user_passwd_match(username, password)
-        if real_username:
+        real_username, fail_reason = userdb.user_passwd_match(username, password)
+        if real_username and fail_reason is None:
             self.logger.info("User %s logged in.", real_username)
             self.do_login(real_username)
         else:
-            self.logger.warning("Failed login for user %s.", username)
-            self.send_message("login_fail")
+            if fail_reason:
+                self.logger.warning("Failed login for user %s: %s", real_username, fail_reason)
```

Comment:
Python supports `elif` as else if, so you don't need an indented `if` block here.